# Φυσικοί αριθμοί

## Οι αριθμοί και η Python

Οι φυσικοί αριθμοί είναι οι αριθμοί από 0, 1, 2, 3, 4, 5, 6, ..., 98, 99, 100, ..., 1999, 2000, 2001, ... μέχρι το άπειρο.

Η Python μπορεί να χειριστεί φυσικούς αριθμούς. Δοκιμάστε στο κουτί παρακάτω έναν φυσικό αριθμό και θα δείτε ότι η Python θα τον επαναλάβει. Π.χ. δείτε τον αριθμό εκατόν είκοσι τρια (123).

In [95]:
123

123


Στην Python όμως θα πρέπει να ακολουθείς κάποιους επιπλέον κανόνες. Για παράδειγμα στους αριθμούς δεν πρέπει να βάζεις τελείες στις χιλιάδες όπως στο χαρτί. 

Αν το κάνεις στην καλύτερη περίπτωση θα προκύψει κάποιο λάθος (αν πατήσεις Εκτέλεση στο παρακάτω θα δεις ότι προκύπτει Syntax Error)

In [96]:
1.000.000

SyntaxError: invalid syntax (1797962719.py, line 1)


Τέτοια λάθη διορθώνονται, όμς με την τελεία μπορεί να συμβεί κάτι χειρότερο. Ο υπολογιστής να καταλάβει διαφορετικό αριθμό από αυτόν που εννοείς.
Δες ποιον αριθμό θα επαναλάβει η Python αν πατήσεις εκτέλεση στο παρακάτω:

In [ ]:
100.000

100.0

Σε αυτό το παράδειγμα, η Python μεταφράζει το 100.000 σε 100.0, που για την Python σημαίνει 100 (εκατό). Γι' αυτόν τον λόγο δεν βάζουμε καθόλου τελείες για τους φυσικούς αριθμούς. Έτσι αν θέλουμε να γράψουμε το ένα εκατομμύριο θα γράψουμε 1000000.

In [ ]:
1000000

1000000


## Πρόσθεση, αφαίρεση και πολλαπλασιασμός φυσικών αριθμών
Μια γλώσσα προγραμματισμού μπορεί να εκτελέσει απλές πράξεις γρήγορα. Στο βιβλίο των μαθηματικών  σου μπορείς να βρεις πολλές ασκήσεις με πράξεις. Μπορείς να τις λύσεις με την Python.

> **ΑΣΚΗΣΗ:** 
> Σελίδα 16
> Να υπολογιστούν τα γινόμενα: 
> 
> (α) $35 \cdot 10$, 
> 
> (β) $421 \cdot 100$,
> 
> (γ) $5 \cdot 1.000$,
> 
> (δ) $27 \cdot 10.000$

Η python μπορεί να κάνει αυτές τις πράξεις ως εξής:

In [ ]:
35*10

350

In [ ]:
421*100

42100

In [ ]:
5*1000

5000

In [ ]:
27*10000

270000


Ο τελεστής του πολλαπλασιασμού είναι το αστεράκι * (SHIFT+8) στο πληκτρολόγιο. Εναλλακτικά, μπορείτε να το βρείτε στο αριθμητικό πληκτρολόγιο. 

> **ΑΣΚΗΣΗ:** 
> (Σελίδα 16)
> Να εκτελεστούν οι ακόλουθες πράξεις:
> 
> (α) $89\cdot 7 + 89\cdot 3$
> 
> (β) $23 \cdot 49 + 77 \cdot 49$
> 
> (γ) $76 \cdot 13 – 76 \cdot 3$
> 
> (δ) $284 \cdot 99$

In [ ]:
89*7+89*3

28116

In [ ]:
23*49+77*49

4900

In [ ]:
76*13-76*3

760

In [97]:
284*99

28116


Στις παραπάνω περιπτώσεις η python εκτελεί πρώτα τους πολλαπλασιασμούς και μετά τις προσθέσεις/αφαιρέσεις δίνοντας έτσι το αποτέλεσμα που αναμένεται. Για παράδειγμα 89\*7 + 89\*3 = 623 + 267 = 890, που είναι το σωστό αποτέλεσμα.
Βέβαια και εσύ με το μυαλό σου κάνεις αυτή την πράξη πολύ γρήγορα γιατί σκέφτεσαι με κοινό παράγοντα το 89, και κάνεις $89(7+3)=89*10=890$ όμως εδώ είναι η δύναμη του υπολογιστή μπορεί να υπολογίζει αυτές τις εκφράσεις με οποιουδήποτε αριθμούς. Δοκίμασε τα παραπάνω κουτάκια κώδικα αλλάζοντας τους αριθμούς για να πειστείς.




> **ΑΣΚΗΣΗ:** 
> (Σελίδα 18)
> Υπολογίστε:
> 
> (α)  $157 + 33$ 
> 
> (β)  $122 + 25 + 78$
> 
> (γ)  $785 - 323$
> 
> (δ)  $7.321 - 4.595$
> 
> (ε)  $60 - (18 - 2)$
> 
> (στ) $52 - 11 -9$
> 
> (ζ)  $23 \cdot 10$
> 
> (η)  $97 \cdot 100$
> 
> (θ)  $879 \cdot 1.000$
Σε python τα παραπάνω υπολογίζονται ως εξής:

In [ ]:
157+33

879000

In [ ]:
122+25+78

In [ ]:
785-323

In [ ]:
7321-4595

In [ ]:
60-(18-2)

In [ ]:
52-11-9

In [ ]:
23*10

In [ ]:
97*100

In [ ]:
879*1000

Οι παρενθέσεις (SHIFT+9 και SHIFT+0) αλλάζουν τη σειρά των πράξεων. Οι πράξεις που είναι μέσα στην παρένθεση εκτελούνται πρώτες. Γι' αυτό το λόγο 60-(18-2)=60-16=44.

> **ΑΣΚΗΣΗ:** 
> (Σελίδα 18)
> Σε ένα αρτοποιείο έφτιαξαν μία μέρα 120 κιλά άσπρο ψωμί, 135 κιλά χωριάτικο, 25 κιλά σικάλεως και 38 κιλά πολύσπορο. Πουλήθηκαν 107 κιλά άσπρο ψωμί, 112 κιλά χωριάτικο, 19 κιλά σικάλεως και 23 κιλά πολύσπορο. Πόσα κιλά ψωμί έμειναν απούλητα;
Με τις γνώσεις που έχουμε θα πρέπει να μετατρέψουμε το παραπάνω πρόβλημα σε μια αριθμητική παράσταση ώστε η python να μπορεί να την υπολογίσει, στη συγκεκριμένη περίπτωση η σωστή παράσταση είναι $$(120-107)+(135-112)+(25-19)+(38-23)$$

In [99]:
(120-107)+(135-112)+(25-19)+(38-23)

57

και η απάντηση είναι 57 κιλά ψωμί.

## Δυνάμεις φυσικών αριθμών
Ο τελεστής της python για τις δυνάμεις είναι ο **  (δυο φορές το αστεράκι). Δηλαδή, αν θέλουμε να υπολογίσουμε το $10^2$ θα γράψουμε 10**2, με όμοιο τρόπο μπορούμε να υπολογίσουμε και τις υπόλοιπες δυνάμεις. Δοκίμασε τα παρακάτω στο REPL.

In [ ]:
10**2

1000000

In [ ]:
10**3

In [ ]:
10**4

Δοκίμασε κι άλλες δυνάμεις αλλάζοντας τον αριθμό που ακολουθεί το **

In [ ]:
10**

Στη προτεραιότητα των πράξεων, οι δυνάμεις έχουν μεγλύτερη προτεραιότητα από τον πολλαπλασιασμό και την πρόσθεση. Οπότε όταν έχουμε και δυνάμεις σε μια παράσταση πρώτα γίνονται οι πράξεις στις παρενθέσεις, μετά οι δυνάμεις και μετά οι πολλαπλασιασμοί και οι προσθέσεις. Την ίδια σειρά ακολουθεί και η python για τον υπολογισμό των πράξεων.
> **ΑΣΚΗΣΗ:** 
> Σελίδα 21
> Να εκτελεστούν οι πράξεις 
> 
>  1. $(2\cdot 5)^4+4\cdot (3+2)^2$
> 
>  2. $(2+3)^3 - 8\cdot 3^2$
> 
Οι αντίστοιχες εκφράσεις είναι (2*5)**4+4*(3+2)**2 



In [ ]:
(2*5)**4+4*(3+2)**2

53

και (2+3)**3 - 8*3**2.

In [ ]:
(2+3)**3 - 8*3**2

H 8*3**2 υπολογίζεται ως $8\cdot (3^2)$, δηλαδή $8\cdot 9 = 72$, αφού πρώτα γίνεται η δύναμη και μετά οι πολλαπλασιασμοί.

> **ΑΣΚΗΣΗ:** 
> Κάνε τις πράξεις: 
> (α) $3\cdot 5^2$, 
> 
> (β) $3\cdot 5^2 + 2$, 
> 
> (γ) $3\cdot5^2 + 2^2$, 
> 
> (δ) $3\cdot 5 + 2^2$, 
> 
> (ε) $3\cdot(5 + 2)^2$.

Αυτές οι πράξεις μπορούν να γίνουν ως εξής:

In [ ]:
3*5**2

147

In [ ]:
3*5**2 + 2

In [ ]:
3*5**2 + 2**2

In [ ]:
3*5 +2**2

In [ ]:
3*(5 + 2)**2


> **ΑΣΚΗΣΗ:** 
> Κάνε τις πράξεις: 
>
> (α) $3^2 +3^3 +2^3 +2^4$, 



In [ ]:
3**2 +3**3 +2**3 +2**4

14686

(β) $(13-2)^ 4 + 5\cdot 3^2$

In [ ]:
(13-2)**4 + 5*3**2


> **ΑΣΚΗΣΗ:** 
> Βρες τις τιμές των παραστάσεων: 
> 
> (α) $(6+5)^2$ και $6^2+5^2$, 
> 
> (β) $(3+6)^2$ και $3^2+6^2$.

In [ ]:
(6+5)**2

45

In [ ]:
6**2+5**2

In [ ]:
(3+6)**2

In [ ]:
3**2+6**2



## Συγκρίσεις φυσικών αριθμών
Μπορούμε να συγκρίνουμε αριθμούς στην Python χρησιμοποιώντας τους τελεστές == (πληκτρολογούμε δύο φορές το =) για την **ισότητα**, > για το **μεγαλύτερο** και < για το **μικρότερο**. Επίσης μπορούμε να χρησιμοποιήσουμε >= για το **μεγαλύτερο ή ίσο** και <= για το **μικρότερο ή ίσο**, τέλος υπάρχει το != για το **δεν είναι ίσο**. 
Η Python επιστρέφει True (αληθές) όταν μία πρόταση ισχύει και False (ψευδές) όταν δεν ισχύει.
Μπορείς να δοκιμάσεις τα παρακάτω:

In [ ]:
123==123

False

In [ ]:
123>123

In [ ]:
123>122

In [ ]:
123<123

In [ ]:
123<124

In [ ]:
123<=123

In [ ]:
123<=124

In [ ]:
123<=122

In [ ]:
123>=123

In [ ]:
123>=124

In [ ]:
123>=122

In [ ]:
122 != 123

In [ ]:
122 != 122

Σκέψου ότι για την Python η σύγκριση είναι και αυτή μια πράξη. Αντί η πράξη αυτή να δίνει σαν αποτέλεσμα έναν αριθμό δίνει σαν αποτέλεσμα το αληθές ή το ψευδές.

Για παράδειγμα:
> **ΑΣΚΗΣΗ:** 
> Να συγκρίνετε τα $3^2$ και $2^3$.
Η σύγκριση αυτή μπορεί να γίνει ως εξής (δοκίμασε να αλλάξεις το > σε == ή != για να δεις τι συμβαίνει):

In [106]:
3**2 > 2**3

True

Άρα το $3^2$ είναι μεγαλύτερο από το $2^3$. Θυμήσου ότι το $3^2=9$, ενώ $2^3=8$.
> **ΑΣΚΗΣΗ:** 

## Η εντολή print
Ήρθε η ώρα να γράψεις εντολές στο πάνω παράθυρο, δηλαδή να γράψεις το πρώτο σου πρόγραμμα.  Με βάση όσα ξέρεις προσπάθησε να γράψεις μια πράξη στο πάνω παράθυρο, για παράδειγμα $32+35$. Ύστερα πάτησε το κουμπί της εκτέλεσης (Run). Μπορείς να δεις το αποτέλεσμα στην εικόνα \ref{noprint}.
\begin{figure}
![noprint.png](noprint.png)
\caption{Η εκτέλεση δεν δίνει κάποιο αποτέλεσμα}
\label{noprint}
\end{figure}

Η Python εκτελεί την πράξη $32+35$, και υπολογίζει το αποτέλεσμα. Αν δεν το έκανε και υπήρχε κάποιο πρόβλημα θα εμφάνιζε κάποιο μήνυμα λάθους στο REPL. Το υπολογισμένο αποτέλεσμα δεν εμφανίζεται. Για να εμφανιστεί το αποτέλεσμα πρέπει να χρησιμοποιήσεις την εντολή print (εκτύπωσε). Η εντολή print εκτελείται ως εξής:

In [107]:
print(32+35)

67


Γράφουμε δηλαδή, print ανοίγουμε παρένθεση, γράφουμε αυτό που θέλουμε να εκτυπωθεί και κλείνουμε την παρένθεση. Όταν εκτελέσουμε το πρόγραμμα με την print τότε εμφανίζεται το αποτέλεσμα στο REPL (εικόνα 1).
![με print](./withprint.png)

Μόλις έγραψες το πρώτο σου πρόγραμμα στην Python. Μάλιστα το πρόγραμμά σου κάνει κάτι. Υπολογίζει το αποτέλεσμα της πράξης $32+35$. 

## Απαρίθμηση
Είδαμε ότι η Python μπορεί να κάνει πολύ γρήγορα, πολύπλοκες πράξεις ακόμη και με δυνάμεις, αλλά δεν είδαμε ακόμη τις απλές ασκήσεις που υπάρχουν στις πρώτες σελίδες του βιβλίου. Όπως για παράδειγμα ποιοι είναι οι τρεις προηγούμενοι αριθμοί του 289 και ποιο οι δύο επόμενοι Σελίδα 13.

Τώρα που μάθαμε να γράφουμε προγράμματα σε Python μπορούμε να αντιμετωπίσουμε αυτό το πρόβλημα με το παρακάτω πρόγραμμα:

In [108]:
print(289-3)
print(289-2)
print(289-1)
print(289+1)
print(289+2)

286
287
288
290
291


που δίνει το αποτέλεσμα

```text
286
287
288
290
291
```

Πιο σωστό θα ήταν να γράψουμε ποιοι αριθμοί είναι οι προηγούμενοι και ποιοι οι επόμενοι. Σε αυτή την περίπτωση θα γράψουμε τις παρακάτω εντολές.

In [109]:
print("Οι  προηγούμενοι αριθμοί είναι:")
print(289-3)
print(289-2)
print(289-1)
print("Οι επόμενοι αριθμοί είναι:")
print(289+1)
print(289+2)

Οι  προηγούμενοι αριθμοί είναι:
286
287
288
Οι επόμενοι αριθμοί είναι:
290
291



Για να εμφανίσει η print τις λέξεις που θέλουμε πρέπει να τις βάλουμε μέσα σε εισαγωγικά. Η Python υποστηρίζει είτε μονά εισαγωγικά, είτε διπλά. Αυτά εισάγονται συνήθως με το ίδιο κουμπί του πληκτρολογίου (κοντά στο ENTER), είτε με SHIFT ή χωρίς. Θυμήσου να κλείνεις τα εισαγωγικά με τον ίδιο τρόπο που τα άνοιξες. Στο πρόγραμμα Mu τα εισαγωγικά αυτά δεν φαίνονται όπως σε άλλα πρόγραμματα σαν `Εισαγωγικά' ή ``Εισαγωγικά " ή <<Εισαγωγικά>>, αλλά φαίνονται κάπως πιο απλά και ίδια στο άνοιγμα και το κλείσιμο \lstinline{'Εισαγωγικά'} ή  \lstinline{"Εισαγωγικά"}. 

Αν θέλουμε να αλλάξουμε το 289 και να βάλουμε έναν άλλο αριθμό,π.χ. το 132 θα πρέπει να αντικαταστήσουμε το 289 μέσα σε όλες τις εντολές print με το 132.

In [110]:
print("Οι προηγούμενοι αριθμοί είναι:")
print(132-3)
print(132-2)
print(132-1)
print("Οι επόμενοι αριθμοί είναι:")
print(132+1)
print(132+2)

Οι προηγούμενοι αριθμοί είναι:
129
130
131
Οι επόμενοι αριθμοί είναι:
133
134



Υπάρχει όμως ένας καλύτερος τρόπος, ο τρόπος αυτός είναι να δώσουμε ένα όνομα στον αριθμό μας. Μπορούμε να πούμε ότι το n είναι το όνομα του αριθμού. Αυτό γίνεται με την εντολή \lstinline{n=132}. Τότε το πρόγραμμά μας γίνεται:

In [111]:
n = 132
print("Οι προηγούμενοι αριθμοί είναι:")
print(n-3)
print(n-2)
print(n-1)
print("Οι επόμενοι αριθμοί είναι:")
print(n+1)
print(n+2)

Οι προηγούμενοι αριθμοί είναι:
129
130
131
Οι επόμενοι αριθμοί είναι:
133
134



Μετά την εντολή `n=132` η Python ξέρει ότι το n είναι ένα όνομα για το 132 και μπορεί να κάνει πράξεις με αυτό. Για παράδειγμα `n+1` κάνει τώρα 133.

Αν θέλουμε να κάνουμε τώρα το ίδιο πρόγραμμα αλλά όχι για το 132 αλλά για το 210, χρειάζεται να αλλάξουμε μόνο μία γραμμή και το πρόγραμμά μας να γίνει ως εξής:

In [112]:
n = 210
print("Οι προηγούμενοι αριθμοί είναι:")
print(n-3)
print(n-2)
print(n-1)
print("Οι επόμενοι αριθμοί είναι:")
print(n+1)
print(n+2)

Οι προηγούμενοι αριθμοί είναι:
207
208
209
Οι επόμενοι αριθμοί είναι:
211
212



Στην Python, όταν δίνουμε ένα όνομα σε έναν αριθμό (με τον τελεστή =) τότε δημιουργούμε μια μεταβλητή. Η μεταβλητή έχει ένα όνομα, στην περίπτωσή μας το n, και μια τιμή, στην περίπτωσή μας το 210.

Αν αντί για τους επόμενους δύο αριθμούς θέλαμε τους επόμενους **δέκα** θα γράφαμε ένα πρόγραμμα όπως το παρακάτω:

In [113]:
n = 210
print(n)
print(n+1)
print(n+2)
print(n+3)
print(n+4)
print(n+5)
print(n+6)
print(n+7)
print(n+8)
print(n+9)
print(n+10)

210
211
212
213
214
215
216
217
218
219
220


Το παραπάνω πρόγραμμα εμφανίζει και τον αριθμό μας n, δηλαδή το 210.

Για να μην γράφουμε πολλές εντολές όταν κάνουμε το ίδιο πράγμα χρησιμοποιούμε την εντολή for.
Το πρόγραμμά μας με την for μπορεί να γίνει:

In [114]:
n = 210
for i in 0,1,2,3,4,5,6,7,8,9,10:
    print(n+i)

210
211
212
213
214
215
216
217
218
219
220


Όταν γράψεις την for στην Python θα πρέπει να δηλώσεις ποιες εντολές θα εκτελεστούν πολλές φορές. Αυτή η δήλωση γίνεται βάζοντας αυτές τις εντολές λίγο πιο μέσα χρησιμοποιώντας το πλήκτρο κενό ή το πλήκτρο tab. Μια καλή πρακτική είναι να βάζεις τέσσερα κενά. Έτσι, πριν την εντολή `print(n+i)` βάζεις τέσσερα κενά.
Το πρόγραμμα αυτό σημαίνει πως για το i μέσα στο σύνολο 0, 1, 2, 3, ... 10 και με αυτή τη σειρά εμφάνισε το n+i. Έτσι το αποτέλεσμα είναι το αναμενόμενο

```text
210
211
212
213
214
215
216
217
218
219
220
```

Στην Python υπάρχει ένας πιο εύκολος τρόπος να γράψουμε τους αριθμούς από το 0 έως το 10. Αυτός ο τρόπος είναι η εντολή range και συγκεκριμένα η range(11). Η range(11) φτιάχνει τους αριθμούς από το 0 μέχρι το 10 οι οποίοι είναι σε πλήθος 11. 
Έτσι το πρόγραμμά μας γίνεται:

In [115]:
n = 210
for i in range(11):
    print(n+i)

210
211
212
213
214
215
216
217
218
219
220



Mπορούμε και να μετρήσουμε τους πρώτους 100 αριθμούς ως εξής:

In [116]:
for i in range(100):
    print(i)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99



Σκέψου αν θα δεις τον αριθμό 100 στο αποτέλεσμα του παραπάνω προγράμματος.

Μπορούμε να δούμε αριθμούς εύκολα με την Python αλλά θα χρειαστεί ξεχωριστό πρόγραμμα αν θέλουμε να εμφανίζεται το λεκτικό  για κάθε αριθμό.

Ένα τέτοιο πρόγραμμα είναι το παρακάτω:

In [117]:
print('μηδέν')
print('ένα')
print('δύο')
print('τρία')
print('τέσσερα')
print('πέντε')
print('έξι')
print('εφτά')
print('οχτώ')
print('εννιά')
print('δέκα')
print('έντεκα')
print('δώδεκα')
print('δεκατρία')
print('δεκατέσσερα')
print('δεκαπέντε')
print('δεκαέξι')
print('δεκαεφτά')
print('δεκαοχτώ')
print('δεκαεννιά')

μηδέν
ένα
δύο
τρία
τέσσερα
πέντε
έξι
εφτά
οχτώ
εννιά
δέκα
έντεκα
δώδεκα
δεκατρία
δεκατέσσερα
δεκαπέντε
δεκαέξι
δεκαεφτά
δεκαοχτώ
δεκαεννιά



Το παραπάνω πρόγραμμα μπορεί να γίνει πιο μαζεμένο με τη χρήση λίστας. Μια λίστα μπορεί να περιέχει τα λεκτικά για κάθε αριθμό. Η λίστα στην Python σημειώνεται με τις τετράγωνες αγκύλες [ και ]. Τα στοιχεία της χωρίζονται με κόμμα. Έτσι η λίστα που θέλουμε τώρα είναι η εξής:

In [118]:
lektika = ['μηδέν','ένα','δύο','τρία','τέσσερα','πέντε','έξι','εφτά','οχτώ','εννιά','δέκα','έντεκα','δώδεκα','δεκατρία','δεκατέσσερα','δεκαπέντε','δεκαέξι','δεκαεφτά','δεκαοχτώ','δεκαεννιά']

Χρησιμοποιούμε τις τετράγωνες αγκύλες και τον αριθμό του στοιχείου που θέλουμε να προσπελάσουμε σε μια λίστα. Η αρίθμηση της λίστας ξεκινάει από το 0. Έτσι, στη λίστα που βλέπουμε παραπάνω το lektika[0] θα είναι η λέξη 'μηδέν' (θυμηθείτε τα εισαγωγικά), το lektika[1] θα είναι η λέξη 'ένα' κ.ο.κ.

Αν θέλετε μπορείτε να κάνετε μια μικρή δοκιμή στο REPL.

In [119]:
lektika = ['μηδέν','ένα','δύο']
lektika[0]
lektika[1]
lektika[2]

'δύο'

Με τη χρήση της λίστας μπορούμε να εμφανίσουμε τους αριθμούς με τη σειρά χρησιμοποιώντας την εντολή for.

In [120]:
lektika = ['μηδέν','ένα','δύο','τρία','τέσσερα','πέντε','έξι','εφτά',
'οχτώ','εννιά','δέκα','έντεκα','δώδεκα','δεκατρία',
'δεκατέσσερα','δεκαπέντε','δεκαέξι','δεκαεφτά','δεκαοχτώ',
'δεκαεννιά']
for i in range(20):
    print(lektika[i])

μηδέν
ένα
δύο
τρία
τέσσερα
πέντε
έξι
εφτά
οχτώ
εννιά
δέκα
έντεκα
δώδεκα
δεκατρία
δεκατέσσερα
δεκαπέντε
δεκαέξι
δεκαεφτά
δεκαοχτώ
δεκαεννιά



Όμως παρότι δεν γράφουμε είκοσι φορές την εντολή print πάλι δίνουμε όλα τα ονόματα στο πρόγραμμά μας βάζοντάς τα σε μια λίστα. Μπορούμε να το αποφύγουμε υπολογίζοντας το λεκτικό. Από το δώδεκα και μέτα το λεκτικό ενός αριθμού i είναι το `δεκα' και μετά το λεκτικό του αριθμού i-10. Για παράδειγμα, το δεκαοχτώ είναι το `δεκα' ακολουθούμενο από το λεκτικό του αριθμού που προκύπτει αν αφαιρέσουμε 10 από το 18.

Η Python μπορεί να κάνει πράξεις και με τις λέξεις, η πρόσθεση λέξεων σημαίνει να τις βάλεις δίπλα δίπλα με τη σειρά. Δοκίμασε 

```text
In [1]: 'δεκα' + 'τρία'
Out[1]: 'δεκατρία'
```
Με αυτή την ευκολία το πρόγραμμά μας γίνεται:

In [121]:
lektika = ['μηδέν','ένα','δύο','τρία','τέσσερα','πέντε','έξι','εφτά',
'οχτώ','εννιά','δέκα','έντεκα','δώδεκα','δεκατρία',
'δεκατέσσερα','δεκαπέντε','δεκαέξι','δεκαεφτά','δεκαοχτώ',
'δεκαεννιά']
for i in range(20):
    if i<=12:
        print(lektika[i])
    else:
        print('δεκα'+lektika[i-10])

μηδέν
ένα
δύο
τρία
τέσσερα
πέντε
έξι
εφτά
οχτώ
εννιά
δέκα
έντεκα
δώδεκα
δεκατρία
δεκατέσσερα
δεκαπέντε
δεκαέξι
δεκαεφτά
δεκαοχτώ
δεκαεννιά




Μάλιστα, το πρόγραμμα υπολογίζει τα λεκτικά από το 13 και μετά και δεν χρειάζεται να τα θυμάται. Μπορούμε να τα διαγράψουμε από τη λίστα.

In [122]:
lektika = ['μηδέν','ένα','δύο','τρία','τέσσερα','πέντε','έξι','εφτά',
'οχτώ','εννιά','δέκα','έντεκα','δώδεκα']
for i in range(20):
    if i > 12:
        print('δέκα'+lektika[i-10])
    else:
        print(lektika[i])

μηδέν
ένα
δύο
τρία
τέσσερα
πέντε
έξι
εφτά
οχτώ
εννιά
δέκα
έντεκα
δώδεκα
δέκατρία
δέκατέσσερα
δέκαπέντε
δέκαέξι
δέκαεφτά
δέκαοχτώ
δέκαεννιά



Τώρα μπορούμε να πάμε μέχρι το 29.

In [123]:
lektika = ['μηδέν','ένα','δύο','τρία','τέσσερα','πέντε','έξι','εφτά',
'οχτώ','εννιά','δέκα','έντεκα','δώδεκα']
for i in range(30):
    if i>20:
        print('είκοσι' + lektika[i-20])
    elif i > 12:
        print('δέκα'+lektika[i-10])
    else:
        print(lektika[i])

μηδέν
ένα
δύο
τρία
τέσσερα
πέντε
έξι
εφτά
οχτώ
εννιά
δέκα
έντεκα
δώδεκα
δέκατρία
δέκατέσσερα
δέκαπέντε
δέκαέξι
δέκαεφτά
δέκαοχτώ
δέκαεννιά
δέκαδέκα
είκοσιένα
είκοσιδύο
είκοσιτρία
είκοσιτέσσερα
είκοσιπέντε
είκοσιέξι
είκοσιεφτά
είκοσιοχτώ
είκοσιεννιά


Το elif είναι συντομογραφία για το else if. Στο σημείο που το έβαλες τώρα σημαίνει αν το i δεν είναι μεγαλύτερο του 20 (else) και είναι μεγαλύτερο από το 12 (if). Άρα το \lstinline{print('δέκα'+lektika[i-10])} γίνεται μόνο αν το i είναι μικρότερο ή ίσο του 20 και μεγαλύτερο από 12.
Το αποτέλεσμα φαίνεται παρακάτω:

```text
 ένα  
 δύο 
 τρία 
 τέσσερα 
 πέντε 
 έξι 
 εφτά 
 οχτώ 
 εννιά 
 δέκα 
 έντεκα 
 δώδεκα 
 δέκατρία 
 δέκατέσσερα 
 δέκαπέντε 
 δέκαέξι 
 δέκαεφτά 
 δέκαοχτώ 
 δέκαεννιά 
  (*@\textcolor{blue}{δέκαδέκα}@*) 
 είκοσιένα 
 είκοσιδύο 
 είκοσιτρία 
 είκοσιτέσσερα 
 είκοσιπέντε 
 είκοσιέξι 
 είκοσιεφτά 
 είκοσιοχτώ 
 είκοσιεννιά 
```
Οπότε καταλαβαίνουμε ότι το είκοσι χρειάζεται ειδικό χειρισμό. Με τη χρήση της elif μπορούμε να βάλουμε και ειδικό χειρισμό για το 20.

In [124]:
lektika = ['μηδέν','ένα','δύο','τρία','τέσσερα','πέντε','έξι','εφτά',
'οχτώ','εννιά','δέκα','έντεκα','δώδεκα']
for i in range(30):
    if i>20:
        print('είκοσι' + lektika[i-20])
    elif i==20:
        print('είκοσι')
    elif i > 12:
        print('δέκα'+lektika[i-10])
    else:
        print(lektika[i])

μηδέν
ένα
δύο
τρία
τέσσερα
πέντε
έξι
εφτά
οχτώ
εννιά
δέκα
έντεκα
δώδεκα
δέκατρία
δέκατέσσερα
δέκαπέντε
δέκαέξι
δέκαεφτά
δέκαοχτώ
δέκαεννιά
είκοσι
είκοσιένα
είκοσιδύο
είκοσιτρία
είκοσιτέσσερα
είκοσιπέντε
είκοσιέξι
είκοσιεφτά
είκοσιοχτώ
είκοσιεννιά


```text
 ένα  
 δύο 
 τρία 
  .
  .
  .
 έντεκα 
 δώδεκα 
 δέκατρία 
  .
  .
  .
 δέκαεννιά 
 (*@\textcolor{blue}{είκοσι}@*) 
 είκοσιένα
  .
  .
  .
 είκοσιεννιά 
```


## Στρογγυλοποίηση
Το βιβλίο των Μαθηματικών της Α' Γυμνασίου αναφέρει πως
Για να στρογγυλοποιήσουμε έναν φυσικό αριθμό Σελίδα 12:
* Προσδιορίζουμε την τάξη στην οποία θα γίνει η στρογγυλοποίηση
* Εξετάζουμε το ψηφίο της αμέσως μικρότερης τάξης
* Αν αυτό το ψηφίο είναι μικρότερο του 5 (δηλαδή 0, 1, 2, 3 ή 4) το ψηφίο αυτό και όλα τα ψηφία των υπόλοιπων τάξεων μηδενίζονται.
* Αν είναι μεγαλύτερο ή ίσο του 5 (δηλαδή 5, 6, 7, 8 ή 9) το ψηφίο αυτό και όλα τα ψηφία των υπόλοιπων τάξεων αντικαθίστανται από το 0 και το ψηφίο της τάξης στρογγυλοποίησης αυξάνεται κατά 1.


Ας πούμε ότι θέλουμε να στρογγυλοποιήσουμε τον αριθμό 454.018.512 στα εκατομμύρια. Η απάντηση που περιμένουμε είναι 454 εκατομμύρια.
Για να τα καταφέρουμε θα χρησιμοποιήσουμε την διαίρεση. Όμως στην Python υπάρχουν **δύο** διαιρέσεις μία με το σύμβολο/και μία με το σύμβολο //. Ας δούμε τις διαφορές τους:

In [125]:
x = 454018512
print(x/1000000)
print(x//1000000)

454.018512
454


Η "κανονική" διαίρεση, με τη μία κάθετο /, δίνει το αποτέλεσμα της διαίρεσης με τα δεκαδικά ψηφία. Η <<ακέραια>> διαίρεση δίνει μόνο τον ακέραιο αριθμό. Δεν μπορούμε να πούμε ότι η ακέραια διαίρεση θα μας δώσει την στρογγυλοποίηση γιατί η ακέραια διαίρεση δεν στρογγυλοποιεί τα δεκαδικά ψηφία αλλά τα απορρίπτει εντελώς. Έτσι, ακόμη και αν είχαμε 454918512 κατοίκους η ακέραια διαίρεση θα δώσει 454 αντί για το στρογγυλοποιημένο που είναι 455.

In [126]:
x = 454918512
print(x/1000000)
print(x//1000000)

454.918512
454



Χρειάζεται επομένως να δούμε το ψηφίο της αμέσως χαμηλότερης τάξης το οποίο είναι το πρώτο δεκαδικό της κανονικής διαίρεσης. Για να το απομονώσουμε αφαιρούμε από το αποτέλεσμα της κανονικής διαίρεσης το ακέραιο μέρος.

In [127]:
x = 454018512
x/1000000 - x//1000000

0.018511999999986983

Οπότε τώρα έχουμε δύο ενδεχόμενα αν το αποτέλεσμα αυτής της πράξης είναι μικρότερο από 0.5 όπως παραπάνω τότε το αποτέλεσμα που ψάχνουμε είναι το αποτέλεσμα της ακέραιας διαίρεσης. Αλλιώς πρέπει να προσθέσουμε ένα στο αποτέλεσμα της ακέραιας διαίρεσης.
Αυτό γίνεται με την εντολή if, που σημαίνει στα αγγλικά αν. Για ευκολία μπορούμε να ονομάσουμε d την διαφορά των δύο διαιρέσεων με την εντολή:

In [128]:
d = x/1000000 - x//1000000

Επειδή το πρόγραμμα γίνεται μεγαλύτερο τώρα θα το γράψουμε στο πάνω παράθυρο του Mu.

In [129]:
x = 454018512
d = x/1000000 - x//1000000
if d < 0.5:
    print(x//1000000)
else:
    print(x//1000000 + 1)

454


Την `if` την γράφουμε ως εξής:

```
if συνθήκη:
    εντολές που εκτελούνται
    αν ισχύει η συνθήκη
else:
    εντολές που εκτελούνται
    αν δεν ισχύει η συνθήκη

```

Θυμήσου να βάζεις την άνω κάτω τελεία μετά τη συνθήκη και μετά τη λέξη else που σημαίνει αλλιώς.

Αν στο ίδιο πρόγραμμα και βάλεις αντί για 454.018.512 τον αριθμό 454.918.512 θα δεις ότι θα εμφανιστεί το σωστό αποτέλεσμα (455).

Αν θέλεις στρογγυλοποίηση στις χιλιάδες τότε το πρόγραμμά σου γίνεται:

In [ ]:
x = 454018512
d = x/1000 - x//1000
if d < 0.5:
    print(x//1000)
else:
    print(x//1000 + 1)

και το αποτέλεσμα είναι 454019.

> **ΑΣΚΗΣΗ:** 
> Για να γίνει το 454.018.512, 450 εκατομμύρια Σελίδα 12 η στρογγυλοποίηση γίνεται στις δεκάδες των εκατομμυρίων. Μπορείς να γράψεις ένα πρόγραμμα που να στρογγυλοποιεί αριθμούς στις δεκάδες των εκατομμυρίων;

## Επανάληψη στις πράξεις
> **ΑΣΚΗΣΗ:** 
> Συμπλήρωσε τον πίνακα τα τετράγωνα και τους κύβους των αριθμών από το 8 μέχρι το 25 (Σελίδα 22).

In [ ]:
for a in range(8,26):
    print(a**2,end =" ")
print()
print()
for a in range(8,26):
    print(a**3,end=" ")

Το αποτέλεσμα αυτού του προγράμματος είναι:



```text
64 81 100 121 144 169 196 225 256 289 324 361 400 441 484 529 576 625 

512 729 1000 1331 1728 2197 2744 3375 4096 4913 5832 6859 8000 9261 
10648 12167 13824 15625
```

Η εντολή print μπορεί να πάρει περισσότερα από ένα ορίσματα, το πρώτο όρισμα είναι αυτό που θα εμφανίσει. Το δεύτερο όρισμα που δώσαμε είναι το end και το ορίσαμε ίσο με το κενό (\lstinline{end=" "}) που σημαίνει ότι η print όταν εμφανίσει το πρώτο όρισμα δεν θα αλλάξει γραμμή αλλά θα αφήσει ένα κενό. Η εντολή print() αλλάζει απλά γραμμή.
> **ΑΣΚΗΣΗ:** 
> Βρες τα τετράγωνα των αριθμών 10,20,30,40,50,60,70,80 και 90 (Σελίδα 22).
Το πρόγραμμα είναι το εξής:

In [ ]:
for i in range(10,100,10):
    print(i**2,end=',')

και το αποτέλεσμα της εκτέλεσης του προγράμματος είναι

```text
100,400,900,1600,2500,3600,4900,6400,8100,
```

> **ΑΣΚΗΣΗ:** 
> Βρες τους κύβους των αριθμών 10,20,30,40,50



In [ ]:
for i in range(10,60,10):
    print(i**3,end=', ')  

Το αποτέλεσμα της εκτέλεσης είναι:

```text
1000, 8000, 27000, 64000, 125000, 
```

## Ανάπτυγμα
> **ΑΣΚΗΣΗ:** (Σελίδα 21) Να γραφεί το ανάπτυγμα του αριθμόύ 7.604 με χρήση των δυνάμεων του 10.
Η απάντηση είναι $7\cdot 10^3 + 6\cdot 10^2 + 0\cdot 10^1 + 4$.  Με συμβολισμό της Python η απάντηση που περιμένουμε είναι:

```text
7*10**3+6*10**2+0*10+4
```



Ας υποθέσουμε ότι ξέρουμε ότι ο αριθμός είναι τετραψήφιος, πως μπορούμε να βρούμε το ανάπτυγμα του. Ξεκινάμε από το πρώτο ψηφίο. Ποιο είναι το πρώτο ψηφίο; Το πρώτο ψηφίο προκύπτει αν διαιρέσουμε τον αριθμό με το 1000 και κρατήσουμε το ακέραιο μέρος.
Δοκίμασε στο REPL:

```text
In [1]: 7604//1000
Out[1]:7
```
Βρήκες το πρώτο ψηφίο, πώς μπορείς να βρεις το δεύτερο; Ας διαιρέσουμε με το 100.

```text
In [1]:7604//100
Out[1]:76
```
Στην ουσία δεν μπορείς να διαιρέσεις τον αρχικό αριθμό με το 100 αλλά αυτόν που σου μένει αφού αφαιρέσεις το πρώτο ψηφίο που έχεις ήδη βρει δηλαδή το 604.

```text
In [2]:604//100
Out[2]:6
```
Έτσι για τα σωστά βήματα είναι:
* Διαιρείς τον αριθμό με το 1000 και κρατάς το ακέραιο μέρος 
* Αφαιρείς από τον αριθμό τις χιλιάδες που βρήκες

Ας ονομάσουμε τον αριθμό 7604, n (`n=7604`), και το πρώτο ψηφίο, στην περίπτωσή μας τις χιλιάδες, prwto.

In [3]:
n = 7604
prwto = n//1000
print(f"Το πρώτο ψηφίο είναι {prwto}.")
n = n - prwto*1000
print(f"Τα υπόλοιπα ψηφία είναι {n}.")

Το πρώτο ψηφίο είναι 7.
Τα υπόλοιπα ψηφία είναι 604.



Ας δούμε λίγο αυτή την εντολή:

In [ ]:
n = n - prwto*1000

Θυμηθείτε ότι εκείνη τη στιγμή το n έχει την τιμή 7604 και το prwto την τιμή 7. Η παραπάνω εντολή σημαίνει:
* Κάνε τις πράξεις που υπάρχουν δεξιά από το σύμβολο ίσον
* Δώσε σαν τιμή στην μεταβλητή που υπάρχει αριστερά από το σύμβολο ίσον το αποτέλεσμα των πράξεων

Έτσι η Python κάνει πρώτα το `n-prwto*1000` δηλαδή $7604-7*1000 = 604$ και αυτό το αποτέλεσμα το δίνει σαν τιμή στην μεταβλητή που υπάρχει αριστερά από το = δηλαδή στη μεταβλητή n. Έτσι το n τώρα είνι 604. **Προσοχή!** Η τιμή $7604$ δεν υπάρχει στην μεταβλητή n. Με αυτόν τον τρόπο το n έχει πάντα τον αριθμό που χρειάζεσαι για να απομονώσεις το επόμενο ψηφίο του αριθμού.

Έτσι ένα συνολικό πρόγραμμα που μπορείς να γράψεις είναι:

In [4]:
n = 7604
prwto = n//1000
n = n - prwto*1000
deutero = n//100
n = n - deutero*100
trito = n //10
n = n - trito*10
tetarto = n
print(prwto,end='')
print('*1000+',end='')
print(deutero,end='')
print('*100+',end='')
print(trito,end='')
print('*10+',end='')
print(tetarto)

7*1000+6*100+0*10+4


Όταν το εκτελέσεις δίνει το σωστό αποτέλεσμα:
```text
7*1000+6*100+0*10+4
```

Το παραπάνω πρόγραμμα δουλεύει με όλους τους τετραψήφιους αριθμούς, απλά άλλαξε το n σε όποιον αριθμό θέλεις στην αρχή του προγράμματος.

Όμως οι 7 εντολές print δεν είναι ο καλύτερος τρόπος να γράψεις το αποτέλεσμα. Θα ήταν καλύτερα να τυπώσουμε αυτά που πρέπει για κάθε ψηφίο χωριστά ώστε να έχουμε τέσσερις εντολές print ως εξής:

In [ ]:
n = 7604
prwto = n//1000
n = n - prwto*1000
deutero = n//100
n = n - deutero*100
trito = n //10
n = n - trito*10
tetarto = n
print(prwto + '*1000+',end='')
print(deutero + '*100+',end='')
print(trito + '*10+',end='')
print(tetarto)


Εξάλλου όταν έχουμε πρόσθεση με λέξεις η Python τις βάζει δίπλα δίπλα οπότε για το `print(prwto + '*1000+',end='')`
θα περιμέναμε να εμφανιστεί το `7*1000`. 
Άν όμως εκτελέσεις το παραπάνω πρόγραμμα θα προκύψει ένα μήνυμα λάθους.

In [ ]:
Traceback (most recent call last):
  File "a.py", line 9, in <module>
    print(prwto + '*10**3+',end='')
TypeError: unsupported operand type(s) for +: 'int' and 'str'

Αν προσέξουμε λίγο θα δούμε ότι το λάθος αφορά την ένατη γραμμή (line 9) και το λάθος είναι TypeError: unsupported operand type(s) for +: 'int' and 'str'.
Σε μετάφραση από τα αγγλικά το μήνυμα λάθους γράφει:

ΛάθοςΤύπων: μη υποστηριζόμενοι τύποι για το +: 'int' και 'str'

Τι είναι οι τύποι και προκύπτουν λάθη από αυτούς;

Οτιδήποτε χρησιμοποιούμε στην Python έχει τύπο. Μάλιστα μπορούμε να δούμε τον τύπο αυτό με την εντολή type. Έτσι δοκίμασε:

```text
In [1]: type(7)
Out[1]:<class 'int'>
In [2]: type('a')
Out[2]:<class 'str'>
In [2]: type('7')
Out[2]:<class 'str'>
```

Βλέπουμε ότι οι αριθμοί έχουν τύπο `int', θα αγνοήσουμε τη λέξη class προς το παρόν. Ενώ οι λέξεις που έχουν τα εισαγωγικά έχουν τύπο `str'. Το `int' προκύπτει από την αγγλική λέξη `integer' που σημαίνει ακέραιος, και το `str' προκύπτει από την αγγλική λέξη `string' που σημαίνει μια σειρά από γράμματα και αριθμούς. Στα ελληνικά το `string' το μεταφράζουμε ως αλφαριθμητικό.

Το πρόβλημα είναι ότι η Python δεν μπορεί να προσθέσει έναν ακέραιο με ένα αλφαριθμητικό. Γι' αυτό δίνει και το μήνυμα λάθους. Ωστόσο, αυτό το πρόβλημα έχει λύση και είναι η μετατροπή του αριθμού σε αλφαριθμητικό με την εντολή str().
Δοκίμασε:

```text
In [1]: type(7)
Out[1]: int

In [2]: str(7)
Out[2]: '7'

In [3]: type('7')
Out[3]: str
```

Το παραπάνω πρόγραμμα γίνεται λοιπόν:

In [8]:
n = 7604
prwto = n//1000
n = n - prwto*1000
deutero = n//100
n = n - deutero*100
trito = n //10
n = n - trito*10
tetarto = n
print(f"{prwto}*1000",end="+")
print(f"{deutero}*100",end="+")
print(f"{trito}*10",end="+")
print(f"{tetarto}")

7*1000+6*100+0*10+4



Που και πάλι δίνει τη σωστή απάντηση.

Καλύτερα είναι να βάλουμε τα στοιχεία prwto, deutero, trito και tetarto σε μια λίστα που θα την ονομάσουμε psifia και θα έχει τέσσερα στοιχεία. Καλό είναι στην αρχή να αρχικοποιούμε τη λίστα με κάποια τιμή ειδικά στην περίπτωση που ξέρουμε πόσο μεγάλη θα είναι όπως τώρα.



In [ ]:
n = 7604
psifia  = [0,0,0,0]
psifia[0] = n//1000
n = n - psifia[0]*1000
psifia[1] = n//100
n = n - psifia[1]*100
psifia[2] = n //10
n = n - psifia[2]*10
psifia[3] = n
print(str(psifia[0]) + '*1000+',end='')
print(str(psifia[1]) + '*100+',end='')
print(str(psifia[2]) + '*10+',end='')
print(str(psifia[3]))

7*10**3+6*10**2+0*10+4



Φαίνεται ότι κάνουμε τα ίδια πράγματα τρεις φορές όπως βλέπεις εδώ:

In [ ]:
psifia[0] = n//1000
n = n - psifia[0]*1000
psifia[1] = n//100
n = n - psifia[1]*100
psifia[2] = n //10
n = n - psifia[2]*10


Θα προσπαθήσουμε να τα κάνουμε με for όπου το i θα μετράει 0,1,2 έτσι το psifia[0] θα γίνει psifia[i]. Όμως θα πρέπει να υπολογίσουμε το 1000 το 1000 είναι 10**3 και στην επόμενη επανάληψη είναι 10**2 κ.ο.κ. οπότε είναι 10**(3-i).
Άρα οι τρεις παραπάνω εντολές μπορούν να αντικατασταθούν με μία for

In [ ]:
for i in range(3):
    psifia[i] = n//10**(3-i)
    n = n - psifia[i]* 10**(3-i)


Το ίδιο πρέπει να γίνει και με τις τρεις εντολές print:

In [ ]:
print(str(psifia[0]) + '*1000+',end='')
print(str(psifia[1]) + '*100+',end='')
print(str(psifia[2]) + '*10+',end='')
print(str(psifia[3]))

```
#TODO: Fix with f string
```

Θα πρέπει να κάνουμε έναν ιδιαίτερο χειρισμό για τις δυνάμεις όπου θα πρέπει να μειώνονται καθώς συνέχίζουν οι επαναλήψεις οπότε αντί για `*10**3΄ χρειαζόμαστε `*10**(3-i)+' όμως το 3-i είναι αριθμός οπότε για να το γράψουμε στην Python θα χρειαστεί να βάλουμε το str() και να γίνει `'*10**'+str(3-i)+'+'`. Οι παραπάνω τρεις εντολές μπορούν τώρα να γραφούν με μία for ως εξής:

In [13]:
for i in range(3):
    print(f"{psifia[i]}*1{'0'*(3-i)}",end='+')

7*1000+6*100+0*10+

και όλο το πρόγραμμα να γίνει:

In [14]:
n = 7604
psifia  = [0,0,0,0]
for i in range(3):
    psifia[i] = n//10**(3-i)
    n = n - psifia[i]* 10**(3-i)
psifia[3] = n
for i in range(3):
    print(f"{psifia[i]}*1{'0'*(3-i)}",end='+')
print(f"{psifia[3]}")

7*1000+6*100+0*10+4



Πώς μπορεί το πρόγραμμα αυτό να δουλεύει για όλους τους ακέραιους ανεξάρτητα από το μέγεθός τους. Αν μετατρέψουμε τον ακέραιο σε αλφαριθμητικό η Python μπορεί να μας πει πόσο μεγάλος είναι με την εντολή len.

In [ ]:
n = 7604
len(str(n))

Το αποτέλεσμα είναι 4.

In [ ]:
n = 7102234
len(str(n))

Το αποτέλεσμα είναι 7.

Και η αρχικοποίηση του πίνακα μπορεί να γίνει για όσα ψηφία θέλουμε (plithos) με την εντολή `[0]*plithos`:

In [ ]:
plithos = 7
psifia = [0] * plithos

Το αποτέλεσμα είναι [0, 0, 0, 0, 0, 0, 0]


Αν υπολογίσουμε το plithos των ψηφίων με το len(str(n)) θα προκύψει 4, ωστόσο εμείς στο πρόγραμμά μας κάνουμε επαναλήψεις τρεις φορές γιατί χρειαζόμαστε ειδικό χειρισμό στο τελευταίο ψηφίο. Έτσι αντικαθιστούμε το 3 με plithos-1 παντού στο πρόγραμμα.

In [ ]:
n = 7604
plithos = len(str(n))
psifia  = [0]*plithos
for i in range(plithos-1):
    psifia[i] = n//10**(plithos-1-i)
    n = n - psifia[i]* 10**(plithos-1-i)
psifia[plithos-1] = n
for i in range(plithos-1):
    print(str(psifia[i]) + '*10**' + str(plithos-1-i) + '+',end='')
print(str(psifia[plithos-1]))


## Ιστορικό σημείωμα
Σελίδα 17

Όταν ο δάσκαλος ζήτησε από τους υπόλοιπους μαθητές να υπολογίσουν το άθροισμα $1+2+3+...+98+99+100$, πριν οι υπόλοιποι αρχίσουν τις πράξεις ο μικρόος Γκάους το είχε ήδη υπολογίσει. Ο δάσκαλος ρώτησε έκπληκτος πώς το βρήκε. Τότε εκείνος έγραψε στον πίνακα:
O Γκάους σκέφτηκε πως το άθροισμα 
$$
1    +    2 + 3  + ... + 99 + 100
$$
είναι ίδιο με το
$$
100+  99+ 98+ ... +    2  + 1
$$
και αν αθροίσουμε τον πρώτο όρο με τον πρώτο όρο, τον δεύτερο με τον δεύτερο κ.ο.κ. Συνολικά αυτό γίνεται:
$$
\begin{align}
\underbrace{(1+100) + (2+99) + ...  (99+2)+(100+1) }_{50\ \text{φορές}} &=\\
101\cdot 100&\\
\end{align}
$$

Άρα το άθροισμα $1 + 2 + 3+ ... + 99 + 100$ είναι $\frac{1}{2}101\times 100 = 5050$ .

Μπορείς να υπολογίσεις το άθροισμα $1 + 2 + 3+ ... + 999 + 1000$ με τον τρόπο του Γκάους;

Ας δούμε το πρόβλημα από την πλευρά της Python.
Ξέρουμε ήδη να εμφανίζουμε τους αριθμούς από το 1 έως το 100 με το παρακάτω πρόγραμμα:

In [15]:
for i in range(101):
    print(i)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100


Πώς όμως θα αθροίσουμε τους αριθμούς αυτούς. Θα φτάξουμε μιά νέα μεταβλητή athroisma και σε αυτή θα προσθέτουμε το i κάθε φορά. Το πρόγραμμα γίνεται:



In [16]:
athroisma = 0
for i in range(101):
    athroisma = athroisma + i
print(athroisma)

5050



Όμως επειδή το άθροισμα είναι χρήσιμο σε πολλές περιπτώσεις η Python έχει έτοιμο το άθροισμα με την εντολή sum. Δοκίμασε:

In [17]:
sum([1,2,3])

6


Με τον ίδιο τρόπο μπορείς να βρεις το άθροισμα $1 + 2 + 3+ ... + 99 + 100$:

In [18]:
sum(range(101))

5050


Τέλος ο τρόπος του μικρού Γκάους είναι:

In [19]:
101*100/2

5050.0


Με την Python μπορούμε να υπολογίσουμε το άθροισμα από το 1 έως το 1000 και με τους δύο τρόπους:

In [20]:
sum(range(1001))
1000*1001/2

500500.0

Στην Python to 500500.0 σημαίνει πως το δεκαδικό μέρος είναι 0 οπότε το αποτέλεσμα είναι και πάλι 500500.


## Ευκλείδια διαίρεση

> **ΑΣΚΗΣΗ:** Για να αποφασίσει ο καθηγητής με ποιο τρόπο θα παρατάξει τους 168 μαθητές για την παρέλαση, πρέπει να διαιρέσει το 168 με τους αριθμούς 3, 4, 5, 6 και 7. 

Διαίρεση με το τρία:

In [ ]:
print(168/3,168%3)

Η διαίρεση γίνεται με τον τελεστή /, οπότε 168/3 είναι το πηλίκο της διαίρεσης του 168 με το 3. Στους υπολογιστές δεν χρησιμοποιούμε την άνω κάτω τελεία για διαίρεση αλλά την κάθετο /, που συνήθως βρίσκεται χαμηλά στο πληκτρολόγιο και στο αριθμητικό πληκτρολόγιο. 

Το υπόλοιπο δίνεται με τον τελεστή \%, οπότε 168\%3 είναι το υπόλοιπο της διαίρεσης του 168 με το 3. Στο συγκεκριμένο παράδειγμα παρατηρούμε ότι το 168 διαιρείται ακριβώς με το 3 και δίνει πηλίκο 56, οπότε ο καθηγητής μπορεί να παρατάξει τους 168 μαθητές σε 56 τριάδες.

Παρόμοια, η διαίρεση του αριθμού 168 με τους αριθμούς 4, 6, και 7 δίνει τα πηλίκα: 42, 28 και 24 αντίστοιχα. 

In [ ]:
print(168/4,168%4)
print(168/6,168%6)
print(168/7,168%7)

Επομένως, μπορούν να παραταχθούν οι μαθητές σε 42 τετράδες ή 28 εξάδες ή σε 24 επτάδες. 

Τέλος, η διαίρεση του 168 με το 5 δίνει πηλίκο 33 και αφήνει υπόλοιπο 3. 

In [ ]:
print(168/5,168%5)

Άρα, δεν μπορεί ο καθηγητής να παρατάξει τους μαθητές σε πλήρεις πεντάδες.

Εκτός από τον τελεστή / για την διαίρεση υπάρχει και ο τελεστής // ο οποίος όμως δίνει το ακέραιο μέρος του πηλίκου. Οπότε αν γράψουμε 168//5 το αποτέλεσμα θα είναι ο ακέραιος αριθμός 33, που όμως είναι μόνο το ακέραιο μέρος του πηλίκου.

Χρησιμοποιώντας αυτούς τους τελεστές μπορεί να γραφεί μια συνάρτηση που να εμφανίζει την ευκλείδια διαίρεση μεταξύ δύο αριθμών:

In [25]:
def eykleidia(D,d):
  print(f"{D}:{d} ευκλείδια {D}={D//d}*{d}+{D%d}")

eykleidia(169,4)

169:4 ευκλείδια 169=42*4+1


Για παράδειγμα (Άσκηση 1):

In [26]:
eykleidia(4002,69)
eykleidia(1445,17)
eykleidia(925,37)
eykleidia(3621,213)
eykleidia(35280,2940)
eykleidia(5082,77)

4002:69 ευκλείδια 4002=58*69+0
1445:17 ευκλείδια 1445=85*17+0
925:37 ευκλείδια 925=25*37+0
3621:213 ευκλείδια 3621=17*213+0
35280:2940 ευκλείδια 35280=12*2940+0
5082:77 ευκλείδια 5082=66*77+0



Άσκηση 2:

In [27]:
eykleidia(65,5)
eykleidia(30,3)
eykleidia(46592,52)

65:5 ευκλείδια 65=13*5+0
30:3 ευκλείδια 30=10*3+0
46592:52 ευκλείδια 46592=896*52+0


Άσκηση 3:

In [28]:
eykleidia(125,3)
eykleidia(762,19)
eykleidia(1500,35)
eykleidia(300,18)

125:3 ευκλείδια 125=41*3+2
762:19 ευκλείδια 762=40*19+2
1500:35 ευκλείδια 1500=42*35+30
300:18 ευκλείδια 300=16*18+12



> **ΑΣΚΗΣΗ:**  Ποια μπορεί να είναι τα υπόλοιπα της διαίρεσης ν:8. 



In [29]:
for i in range(20):
    print(i%8)

0
1
2
3
4
5
6
7
0
1
2
3
4
5
6
7
0
1
2
3



To υπόλοιπο μπορεί να είναι από 0 έως 7.


> **ΑΣΚΗΣΗ:**  Αν ένας αριθμός διαιρεθεί δια 9 δίνει πηλίκο 73 και υπόλοιπο 4. Ποιος είναι ο αριθμός;



In [ ]:
print(9*73+4)
eykleidia(661,73)


Ο αριθμός αυτός είναι $9\cdot 73+4 = 661$.

> **ΑΣΚΗΣΗ:**  Αν σήμερα είναι Τρίτη, τι μέρα θα είναι μετά από 247 ημέρες;

Για να λυθεί αυτό το πρόβλημα θα κατασκευάσουμε έναν πίνακα meres:

In [ ]:
meres = ['ΤΡ','ΤΕ','ΠΕ','ΠΑ','ΣΑ','ΚΥ','ΔΕ']

Σε αυτόν meres[0] = 'ΤΡ', meres[1]='ΤΕ' κ.λπ.
Αυτό το πρόβλημα προσεγγίζεται ως εξής:
1. Μετά από μία(1) ημέρα θα είναι Τετάρτη(meres[1])
2. Μετά από δυό(2) ημέρες θα είναι Πέμπτη(meres[2])
3. Μετά από τρεις(3) ημέρες θα είναι Παρασκευή(meres[3])
4. Μετά από τέσσερις(4) ημέρες θα είναι Σάββατο(meres[4])
5. Μετά από πέντε(5) ημέρες θα είναι Κυριακή(meres[5])
6. Μετά από έξι(6) ημέρες θα είναι Δευτέρα(meres[6])
7. Μετά από επτά(7) ημέρες θα είναι Τρίτη(meres[0] = meres[7 \% 7]), αφού το υπόλοιπο της διαίρεσης του 7 με το 7 είναι 0.
8. Μετά από οκτώ(8) ημέρες θα είναι Τετάρτη(meres[1] = meres[8 \% 7]), αφού το υπόλοιπο της διαίρεσης του 8 με το 7 είναι 1.

κατά συνέπεια μετά από 247 ημέρες θα είναι Πέμπτη:

In [ ]:
meres = ['ΤΡ','ΤΕ','ΠΕ','ΠΑ','ΣΑ','ΚΥ','ΔΕ']
print(meres[247%7])

## Χαρακτήρες διαιρετότητας - ΜΚΔ - ΕΚΠ - Ανάλυση αριθμού σε γινόμενο πρώτων παραγόντων
> **ΑΣΚΗΣΗ:** Το τοπικό γραφείο της UNICEF θα μοιράσει 150 τετράδια, 90 στυλό και 60 γόμες σε πακέτα δώρων, ώστε τα πακέτα να είναι τα ίδια και να περιέχουν και τα τρία είδη.
> **ΑΣΚΗΣΗ:**  1. Μπορεί να γίνουν 10 πακέτα δώρων; Αν ναι, πόσα από κάθε είδος θα έχει κάθε πακέτο;
> **ΑΣΚΗΣΗ:**  2. Πόσα πακέτα δώρων μπορεί να γίνουν με όλα τα διαθέσιμα είδη;
> **ΑΣΚΗΣΗ:**  3. Πόσα πακέτα δώρων μπορεί να γίνουν με τα λιγότερα δυνατά από κάθε είδος;

1. Μπορούν να γίνουν 10 πακέτα δώρων, με 15 τετράδια, 9 στυλό και 6 γόμες. 

In [ ]:
150%10
90%10
60%10

2. Ακόμη και δύο πακέτα μπορούν να γίνουν με 75 τετράδια, 45 στυλό και 30 γόμες.
3. Για να έχουμε μέσα τα λιγότερα δυνατά είδη θα έχουμε περισσότερα πακέτα και θα πρέπει και οι τρεις αριθμοί να διαιρούν ακριβώς το πλήθος των πακέτων:

In [ ]:
for i in range(1,60):
    if (150 % i == 0 and 90 % i == 0 and 60 % i == 0):
        print(i)

Έτσι η απάντηση είναι 30 πακέτα με 150:30 = 5 τετράδια, 3 στυλό και 2 γόμες το καθένα.

Το μέγιστο πλήθος των πακέτων που διαιρεί και τους τρεις αριθμούς είναι ο μέγιστος κοινός διαιρέτης τους, ή αλλιώς Μ.Κ.Δ., για να βρεθεί ο Μ.Κ.Δ. δύο αριθμών στην Python 3 μπορεί να χρησιμοποιηθεί το παρακάτω πρόγραμμα:

In [ ]:
from fractions import gcd
print(gcd(150,90))

Η συνάρτηση gcd υπολογίζει τον Μ.Κ.Δ. δύο αριθμών.

Για τους τρεις αριθμούς, υπάρχει το εξής πρόβλημα αν τους τοποθετήσετε σαν ορίσματα της συνάρτησης gcd

In [ ]:
print(gcd(150,90,60))

Η φράση *gcd() takes 2 positional arguments but 3 were given* σημαίνει πως δεν μπορούμε να υπολογίσουμε τον Μ.Κ.Δ. βάζοντας όλους τους αριθμούς σαν ορίσματα της συνάρτησης. Ευτυχώς, ο Μ.Κ.Δ. των τριών αριθμών μπορεί να υπολογιστεί ως εξής:



In [ ]:
mkd150_90 = gcd(150,90)
print(gcd(mkd150_90,60))

που είναι το ίδιο με τον παρακάτω κώδικα:

In [ ]:
print(gcd(gcd(150,90),60))


### Ανάλυση σε γινόμενο πρώτων παραγόντων

Το παρακάτω πρόγραμμα αναλύει αριθμούς σε γινόμενο παραγόντων.

In [ ]:
class ginomenoparagontwn():
  def __init__(self,n):
    self.paragontes = []
    self.dinameis = {}
    if type(n) == int:
      self.arithmos = n
      while n <1:
        for i in range(2,n+1):
          if n%i == 0:
            n = n // i
            self.paragontes.append(i)
            break
      self.paragontes = sorted(self.paragontes)
      for i in self.paragontes:
        if i not in self.dinameis:
          self.dinameis[i] = self.paragontes.count(i)
    elif type(n) == dict:
      self.arithmos = 1
      self.dinameis = n
      for i in n:
        self.arithmos *= i**n[i]
      for i in n:
        self.paragontes.extend([i]*n[i])
      self.paragontes = sorted(self.paragontes)
    if len(self.paragontes) < 1:
      self.einaiPrwtos = False
    else:
      self.einaiPrwtos = True
  
  def mkd(self,other):
    if type(other) == int:
      other = ginomenoparagontwn(other)
    dinameismkd = {}
    for i in self.dinameis:
      if i in other.dinameis:
        dinameismkd[i] = min(self.dinameis[i],other.dinameis[i])
    if dinameismkd == {}:
      return(ginomenoparagontwn({1:1}))
    else:
      return(ginomenoparagontwn(dinameismkd))
      
  def ekp(self,other):
    if type(other) == int:
      other = ginomenoparagontwn(other)
    dinameisekp = self.dinameis
    for i in other.dinameis:
      if i in dinameisekp:
        dinameisekp[i] = max(dinameisekp[i],other.dinameis[i])
      else:
        dinameisekp[i] = other.dinameis[i]
    return(ginomenoparagontwn(dinameisekp))
    
  def __repr__(self):
    return(str(self.arithmos) + ':' + '*'.join([str(x) + '^' + str(self.dinameis[x]) 
      for x in self.dinameis]) + ':' + '*'.join([str(x) for x in self.paragontes]) + 
      (',πρώτος' if self.einaiPrwtos else ""))

> **ΑΣΚΗΣΗ:** 
>  Να αναλυθούν οι αριθμοί 2520, 2940, 3780 σε γινόμενο πρώτων παραγόντων. 



In [ ]:
print(ginomenoparagontwn(2520))
print(ginomenoparagontwn(2940))
print(ginomenoparagontwn(3780))

Επομένως, η ανάλυση σε γινόμενο είναι:
$$2520 = 2^3\cdot 3^2\cdot 5\cdot 7$$
$$2940 = 2^2\cdot 3\cdot 5\cdot 7^2$$
$$3780 = 2^2\cdot 3^3\cdot 5\cdot 7$$

Το ελάχιστο κοινό πολλαπλάσιο (Ε.Κ.Π.) υπολογίζεται ως εξής:



In [ ]:
ginomenoparagontwn(2520).ekp(2940).ekp(3780)

επομένως Ε.Κ.Π.(2520,2940,3780) = 52920.
Ομοίως, ο μέγιστος κοινός διαιρέτης (Μ.Κ.Δ.) υπολογίζεται ως εξής:



In [ ]:
ginomenoparagontwn(2520).mkd(2940).mkd(3780)

και Μ.Κ.Δ.(2520,2940,3780) = 420.

> **ΑΣΚΗΣΗ:** 
>  Να βρεθεί αν διαιρούνται οι αριθμοί 12510, 772, 225, 13600 με 2, 3, 4, 5, 8, 9, 10, 25, 100.

Η φιλοσοφία της άσκησης είναι να λυθεί με τα κριτήρια διαιρετότητας ωστόσο, στον υπολογιστή ο υπολογισμός του υπολοίπου είναι εύκολος με τον τελεστή %.

In [ ]:
diairetaios = [12510,772,224,13600]
diairetis = [2,3,4,5,8,9,10,25,100]
print(" "*5,end=',')
for d in diairetis:
    print(str(d).rjust(3),end = ',')
print()
for D in diairetaios:
    print(str(D).rjust(5),end=',')
    for d in diairetis:
        if (D%d==0):
            print('Ναι'.rjust(3),end = ',')
        else:
            print('Όχι'.rjust(3),end = ',')
    print()

Το αποτέλεσμα είναι:

```text
     ,  2,  3,  4,  5,  8,  9, 10, 25,100,
12510,Ναι,Ναι,Όχι,Ναι,Όχι,Ναι,Ναι,Όχι,Όχι,
  772,Ναι,Όχι,Ναι,Όχι,Όχι,Όχι,Όχι,Όχι,Όχι,
  224,Ναι,Όχι,Ναι,Όχι,Ναι,Όχι,Όχι,Όχι,Όχι,
13600,Ναι,Όχι,Ναι,Ναι,Ναι,Όχι,Ναι,Ναι,Ναι,
```

Το άθροισμα των ψηφίων ενός αριθμού μπορεί να υπολογιστεί με την εξής εντολή:

In [ ]:
sum([int(i) for i in str(num)])

Για παράδειγμα

In [ ]:
num = 123
sum([int(i) for i in str(num)])

Αυτή η εντολή ανήκει στην κατηγορία του list comprehension. Η εναλλακτική είναι το εξής πρόγραμμα:

In [ ]:
athroisma = 0
for i in str(num):
    athroisma = athroisma + i

Το αποτέλεσμά του είναι:

In [ ]:
TypeError: unsupported operand type(s) for +: 'int' and 'str'

Ο λόγος είναι σε αυτό το i είναι το κάθε ψηφίο αλλά σαν γράμμα της αλφαβήτου του υπολογιστή (αλφαριθμητικό / string). Έτσι το σωστό πρόγραμμα είναι:

In [ ]:
athroisma = 0
for i in str(num):
    athroisma = athroisma + int(i)

που υπολογίζει το άθροισμα των ψηφίων.

Στην python ο τελεστής sum, υπολογίζει το άθροισμα των στοιχείων μιας λίστας όταν αυτή αποτελείται από αριθμούς για παράδειγμα

In [ ]:
sum([1,2,3])

Επομένως πρέπει να κατασκευάσουμε μια λίστα στην οποία οι αριθμοί θα είναι τα ψηφία του αρχικού αριθμού num. Η κατασκευή αυτής της λίστας μπορεί να γίνει πάλι με μία for και στη συνέχεια να υπολογιστεί το άθροισμα :

In [ ]:
lista = []
for i in str(num):
    lista.append(int(i))
sum(lista)

Με το list comprehension όμως η δημιουργία της λίστας απλοποιείται, γράφοντας το for μέσα στον πίνακα και γίνεται:

In [ ]:
lista = [int(i) for i in str(num)]
sum(lista)

και τέλος

In [ ]:
sum([int(i) for i in str(num)])

Το άθροισμα των ψηφίων των παραπάνω αριθμών μπορεί να υπολογιστεί με το παρακάτω πρόγραμμα:

In [ ]:
diairetaios = [12510,772,224,13600]
for num in diairetaios:
    print(num,sum([int(i) for i in str(num)]))

και είναι

```text
12510 9
772 16
224 8
13600 10
```
\subsubsection{Ασκήσεις}
> **ΑΣΚΗΣΗ:** 
>  1. Ισχύει ότι: (100 - 30) - 10 = 100 - (30 - 10)



In [ ]:
(100 - 30) - 10 == 100 - (30 - 10)

         
> **ΑΣΚΗΣΗ:** 
>  2. Για να πολλαπλασιάσουμε έναν αριθμό με το 11 τον πολλαπλασιάζουμε με το 10 και προσθέτουμε 1.

Ισχύει   
> **ΑΣΚΗΣΗ:** 
>  3. To γινόμενο $3\cdot 3 \cdot 3$ γράφεται $3^3$.



In [ ]:
3*3*3 == 3**3

> **ΑΣΚΗΣΗ:** 
>  4.     Το $2^5$ ισούται με 10.



In [ ]:
2**5 == 10

> **ΑΣΚΗΣΗ:** 
>  5. $\alpha + \alpha + \alpha + \alpha = 4 \cdot \alpha$.

Ισχύει
> **ΑΣΚΗΣΗ:** 
>  6. $\alpha\cdot \alpha\cdot \alpha\cdot \alpha\cdot \alpha\cdot \alpha = \alpha^5$.

Δεν ισχύει
> **ΑΣΚΗΣΗ:** 
>  7. $2^3 + 3 = 11$.



In [ ]:
2**3 + 3 == 11

 > **ΑΣΚΗΣΗ:** 
> 8. $3\cdot 10^2 + 2 \cdot 10^1 + 2 \cdot 10^0 = 322$.



In [ ]:
3*10**2 + 2*10**1 + 2*10**0 == 322

> **ΑΣΚΗΣΗ:** 
>  9.     $20 - 12 : 4 = 2$.



In [ ]:
20-12/4 == 2

> **ΑΣΚΗΣΗ:** 
>  10. $9 \cdot 3 - 2 + 5 = 30$.



In [ ]:
9*3-2+5 == 11

> **ΑΣΚΗΣΗ:** 
>  11. $(3 \cdot 1 - 3) : 3 = 0$.



In [ ]:
(3*1-3)/3 == 0


> **ΑΣΚΗΣΗ:** 
>  12. Στη σειρά των πράξεων:  $7 + (6 \cdot 5) + 4$, οι παρενθέσεις δεν χρειάζονται.



In [ ]:
7+(6*5)+4 == 7+6*5+4


> **ΑΣΚΗΣΗ:** 
>  13. Η διαφορά δύο περιττών αριθμών είναι πάντα περιττός αριθμός.

Δεν ισχύει
> **ΑΣΚΗΣΗ:** 
>  14. Αν ο αριθμός α είναι πολλαπλάσιο του αριθμού β, τότε ο α διαιρείται με το β.

Ισχύει   
> **ΑΣΚΗΣΗ:** 
>  15.    Το 38 είναι πολλαπλάσιο του 2 και του 3.



In [ ]:
38%2
38%3

Το 38 δεν είναι πολλαπλάσιο του 3

> **ΑΣΚΗΣΗ:** 
>  16. Ο αριθμός 450 διαιρείται με το 3 και το 9.



In [ ]:
450%3
450%9

Άρα, ο αριθμός 450 διαιρείται και με το 3 και με το 9.

> **ΑΣΚΗΣΗ:** 
>  17. Ο 35 και ο 210 έχουν μέγιστο κοινό διαιρέτη τον αριθμό 5.



In [ ]:
ginomenoparagontwn(35).mkd(210)

Δεν ισχύει, έχουν το 35.
> **ΑΣΚΗΣΗ:** 
>  18. Το ΕΚΠ των 2 και 24 είναι ο αριθμός 48.



```text
ginomenoparagontwn(2).mkd(24)
24:2^3*3^1:2*2*2*3
```
> **ΑΣΚΗΣΗ:** 
>  19. Η διαίρεση 420 : 15 δίνει υπόλοιπο 18.



In [ ]:
420 % 15

Δεν ισχύει
> **ΑΣΚΗΣΗ:** 
>  20. Η σχέση 177 = 5 · 35 + 2 είναι μια ευκλείδια διαίρεση.



In [ ]:
177 == 5 * 35 + 2

και επίσης, το υπόλοιπο (2) είναι μικρότερο του διαιρέτη 5.      

> **ΑΣΚΗΣΗ:** 
>  21. Ο αριθμός 3 · α + 9 διαιρείται με το 3.

$$\frac{3\cdot \alpha +9}{3} = \alpha + 3$$
Οπότε αν ο $\alpha$ είναι ακέραιος τότε το $3\cdot \alpha +9$ διαρείται με το 3.
> **ΑΣΚΗΣΗ:** 
>  22. Ο αριθμός $300$ αναλύεται σε γινόμενο πρώτων παραγόντων ως $3\cdot 10^2$.



In [ ]:
ginomenoparagontwn(300)

> **ΑΣΚΗΣΗ:** 
>  23. Ο αριθμός 224 διαιρείται με το 4 και το 8.



In [ ]:
224%4
224%8

Ισχύει!